[CORL-2370] Add archiving copy to stream/admin

marcushaddon opened the pull request:
## What does this PR do?

- Injects the `archivingEnabled` and `autoArchiveOlderThanMs` into the `staticConfig` and from there into the `Local` relay store for both admin and stream
- If archiving is enabled, shows the following throughout stream/admin
    - `My Profile > my comments` will indicate that the comments shown are only those that are new enough to not be archived based on the archiving age threshold
    - Dashboard labels will show that they are relative to the archiving threshold and no longer for "all time"
    - User history dashboard will indicate that the comments shown are only those that are new enough to not be archived based on the archiving age threshold

## What changes to the GraphQL/Database Schema does this PR introduce?

None

## How do I test this PR?

- Enable archiving by setting `MONGODB_ARCHIVE_URI` in your `.env` file
    - I use a second mongo instance using this command: 
     ```
     docker run -d -p 27018:27017 --restart always --name mongo-archive mongo:4.2
     ```
    - I then use the connection string for that second Mongo I spun up for archiving in my `.env` file:
     ```
     MONGODB_ARCHIVE_URI=mongodb://127.0.0.1:27018/coral
     ```
    - The reason I spin up another Mongo is because archiving is meant to be used with archived comments living in a different cluster (mongo instance) than the live comments. This docker setup matches production most closely for local testing.
- Build and run using `npm run build:development` and `npm run start:development`
- Use an external commenting/iframe host (i.e. the `multisite-test-bed`)
    - This is because the `npm run watch` tooling does not properly update the local state and staticConfig injection due to its configuration
- Check that `My Profile > my comments`, the dashboard, and the user history all comments tab show the appropriate archiving specific notifications and labels
    - Further details on what that should be are listed above in `What does this PR do?`
 
## How do we deploy this PR?

No special considerations should be needed.

marcushaddon committed:
WIP add archiving copy

marcushaddon committed:
rename archiving static config vars

marcushaddon committed:
remove various debugging cruft

marcushaddon committed:
create millisToDays helper
create localizations

nick-funk committed:
Revert "create millisToDays helper"

This reverts commit 692610f5fae58e003e16537cd2bf79e5242d8180.

nick-funk committed:
Revert "remove various debugging cruft"

This reverts commit dcdcfc29ec6cfb68f07f6cc9dc0715d2c4cc7e4a.

nick-funk committed:
Revert "rename archiving static config vars"

This reverts commit 7a5d172042fcd49cda18cd8622ccf48af5fba01d.

nick-funk committed:
Revert "WIP add archiving copy"

This reverts commit 5ab5cf1d582cf7ac92af55320ed1ed3bc0ddea20.

nick-funk committed:
Inject archiving variables into static config

nick-funk committed:
Show archive threshold notification under My Profile > My comments

nick-funk committed:
Add archived threshold notification to the user history drawer

nick-funk committed:
Add proper spacing and styling to archive threshold notifications

nick-funk committed:
Update dashboard total labels when archiving is enabled

nick-funk committed:
Update testing initLocalState with archiving staticConfig values

nick-funk committed:
Use new archivingEnabled local variable to hide/show archive story actions

nick-funk committed:
Don't pass a non-existent parameter to component

nick-funk committed:
Fix incorrect import

nick-funk committed:
Add tests for conditional showing of My Comments archived notification

kabeaty commented on src/core/client/stream/test/create.tsx:
```diff
@@ -26,6 +27,16 @@ export default function create(params: CreateTestRendererParams) {
+        DEFAULT_AUTO_ARCHIVE_OLDER_THAN,
+        "autoArchiveOlderThanMs"
+      );
+      localRecord.setValue(false, "archivingEnabled");
```
Looks like these values are being set twice?

nick-funk replied:
Hah! Yup, fixed!

kabeaty commented on src/core/client/stream/test/profile/myComments.spec.tsx:
```diff
@@ -115,3 +127,27 @@ it("loads more comments", async () => {
+});
+
+it("doesn't show archived notification when archiving disabled", async () => {
+  const testRenderer = createTestRenderer({ archivingEnabled: false });
```
We don't need to set `archivingEnabled` to false here, right, since that's the default?

nick-funk replied:
I like to be explicit when I do tests. Even though it's the default (and I set the default to make future tests easier to setup if you don't need to be explicit). For this test, it is important that it's false, so setting it to false explicitly conveys what the test is testing.

That's just a personal preference of mine. Verbosity is preferable in my opinion when testing, even when it seems redundant.

kabeaty commented on src/core/common/config.ts:
```diff
@@ -57,4 +57,16 @@ export interface StaticConfig {
   forceAdminLocalAuth: boolean;
+
+  /**
+   * archivingEnabled will be true when the deployment has set a valid Mongo DB
```
nit - Do you mind updating this to `MongoDB` instead of `Mongo DB`? Will help with searching to have correct/consistent.

nick-funk replied:
Yup! Fixed!

kabeaty commented on src/core/common/config.ts:
```diff
@@ -57,4 +57,16 @@ export interface StaticConfig {
+  archivingEnabled: boolean;
+
+  /**
+   * autoArchivingIntervalMs is the time in milliseconds that a story will
```
Can you update this comment to say `autoArchiveOlderThanMs` to match what it's for?

nick-funk replied:
Yup! Fixed!

kabeaty commented on src/core/client/admin/routes/Dashboard/sections/Today.tsx:
```diff
@@ -30,6 +36,22 @@ interface Props {
+    }
+  `);
+
+  const seconds = Math.floor(autoArchiveOlderThanMs / 1000);
```
Since this and the below is reused in a few places, does it make sense to add it to a shared helper somewhere?

nick-funk replied:
I generally agree with grouping similar functionality into a shared function. Removes duplicates of the same code. In this case though, I started thinking "what would I name this?" and got `convertMillisecondsToSeconds` and got flashbacks of that time I was 12 and I tried to re-invent DateTime utility functions. Given the specificity for the flooring that makes the rounding not re-usable for accuracy and the simplicity of just dividing by 1000, it seems un-necessary.

I wish I could just put the localization for this into a re-usable component and just use it for both of the notifications, but FTL dislikes it if you try and nest a `<Localized>` in another `<Localized>`... I'll grump about that later.

kabeaty commented on src/core/client/stream/test/profile/myComments.spec.tsx:
```diff
@@ -115,3 +127,27 @@ it("loads more comments", async () => {
+
+  expect(
+    within(commentHistory).getByTestID("archivedComments-thresholdNotification")
+  ).toBeDefined();
```
What if we get this by the actual text that we expect it to show to check that's as expected? I also think that would be more readable.

nick-funk replied:
Can do, but I won't be able to do the whole text, since it's interpolated, but I'll do the first bit that is consistent and searchable. Should get it towards more read-able.

kabeaty commented on src/locales/en-US/admin.ftl:
```diff
@@ -1541,10 +1574,103 @@ dashboard-heading-last-updated = Last updated:
+  } average
 dashboard-today-staff-comments = Staff comments
 dashboard-alltime-staff-comments = All time total
+dashboard-alltime-staff-comments-archiveEnabled = { $value } { $unit ->
```
Is it possible to reuse the `framework-timeago` that already exists for the places where we are using this in translations? And then a separate translation for the `average`/`total`, etc.?

nick-funk replied:
I really want to do this! I've tried to nest Localizations in the past, FTL doesn't seem to allow it. It's kind of annoying and I wish I could do it, but I've never successfully managed to do so.

kabeaty replied:
Ah...hmm. Would it work to do something like two `span`s beside each other, since that wouldn't be nesting the localizations?

nick-funk replied:
I agree with your attempt to do what we can to make modular translation components. I really want to do that!

I try not to do multi-span/div translations for when someone localizes based on a language that goes right to left instead of left to right. I want them to be able to localize it seamlessly as one `<Localized>` object that they just update the text + values for.

kabeaty reviewed (approved):
Looks good, and works as expected! Approving, with some small comments/questions and potential ideas for refactoring in a few places.

nick-funk committed:
Remove duplicate set of autoArchive local var's in stream test create

nick-funk committed:
Rename `Mongo DB` to `MongoDB`

nick-funk committed:
Update out of date config comment

nick-funk committed:
Update test to search by actual text used in UI

Instad of a test-id, use the text we use in the UI. This will
have to be updated if the text changes, but is more read-able.

kodiakhq[bot] committed:
Merge branch 'develop' into feat/CORL-2370-add-archiving-copy-text
